src/utils/window.rs

```diff
@@ -16,39 +16,82 @@ use windows::Win32::{
             PROCESS_VM_READ,
         },
     },
-    UI::WindowsAndMessaging::{
-        EnumWindows, GetCursorPos, GetForegroundWindow, GetWindow, GetWindowLongPtrW,
-        GetWindowPlacement, GetWindowTextW, GetWindowThreadProcessId, IsIconic, IsWindowVisible,
-        SetForegroundWindow, SetWindowPos, ShowWindow, GWL_EXSTYLE, GWL_USERDATA, GW_OWNER,
-        SWP_NOZORDER, SW_RESTORE, WINDOWPLACEMENT, WS_EX_TOPMOST,
+    UI::{
+        Controls::STATE_SYSTEM_INVISIBLE,
+        WindowsAndMessaging::{
+            EnumWindows, GetCursorPos, GetForegroundWindow, GetTitleBarInfo, GetWindow,
+            GetWindowLongPtrW, GetWindowPlacement, GetWindowTextW, GetWindowThreadProcessId,
+            IsIconic, IsWindowVisible, SetForegroundWindow, SetWindowPos, ShowWindow, GWL_EXSTYLE,
+            GWL_USERDATA, GW_OWNER, SWP_NOZORDER, SW_RESTORE, TITLEBARINFO, WINDOWPLACEMENT,
+            WS_EX_TOPMOST,
+        },
     },
 };
 
+use crate::utils::RegKey;
+
 pub fn is_iconic_window(hwnd: HWND) -> bool {
     unsafe { IsIconic(hwnd) }.as_bool()
 }
 
 pub fn is_visible_window(hwnd: HWND) -> bool {
     let ret = unsafe { IsWindowVisible(hwnd) };
-    ret.as_bool()
+    if !ret.as_bool() {
+        return false;
+    }
+
+    // Some "visible" windows are cloaked with `DWM_CLOAKED_SHELL` but always have
+    // this invisible flag set, so this filters out those unusual cases:
+    let mut title_info = TITLEBARINFO::default();
+    title_info.cbSize = std::mem::size_of_val(&title_info) as u32;
+    let _ = unsafe { GetTitleBarInfo(hwnd, &mut title_info) };
+
+    title_info.rgstate[0] & STATE_SYSTEM_INVISIBLE.0 == 0
 }
 
 pub fn is_topmost_window(hwnd: HWND) -> bool {
     let ex_style = unsafe { GetWindowLongPtrW(hwnd, GWL_EXSTYLE) } as u32;
     ex_style & WS_EX_TOPMOST.0 != 0
 }
 
-pub fn is_cloaked_window(hwnd: HWND) -> bool {
-    let mut cloaked = 0u32;
+pub fn get_window_cloak_type(hwnd: HWND) -> u32 {
+    let mut cloak_type = 0u32;
     let _ = unsafe {
         DwmGetWindowAttribute(
             hwnd,
             DWMWA_CLOAKED,
-            &mut cloaked as *mut u32 as *mut c_void,
+            &mut cloak_type as *mut u32 as *mut c_void,
             size_of::<u32>() as u32,
         )
     };
-    cloaked != 0
+    cloak_type
+}
+
+fn is_cloaked_window(hwnd: HWND, only_current_desktop: Option<bool>) -> bool {
+    let only_current_desktop = only_current_desktop.unwrap_or_else(|| {
+        let alt_tab_filter = RegKey::new_hkcu(
+            w!(r"Software\Microsoft\Windows\CurrentVersion\Explorer\Advanced"),
+            w!("VirtualDesktopAltTabFilter"),
+        )
+        .and_then(|k| k.get_int())
+        .unwrap_or(1);
```

Comment:
Oh, I forgot, I was wondering if it made any sense to try and cache this on startup along with the other configs.

I assume registry lookups are fairly efficient, and this way allows users to change it on the fly (like Windows), but it does mean this reg key is read every time the user hits their hotkey....

Reply:
The registry is cached in memory, so it's ok to read it every time the user triggers a switch.